content/5_Plat/RURQ.mdx

```diff
@@ -152,135 +152,135 @@ Given the mechanics behind the `pushdown` function, all you need is a regular
 range-sum segment tree to solve the question.
 
 ```cpp
-
-#include <iostream>
-using ll = long long;
+#include <bits/stdc++.h>
 using namespace std;
+using ll = long long;
 
-const int maxN = 2e5 + 5;
-
-int N, Q;
-int a[maxN];
+/**
+ * Represents the type of lazy update being done.
+ * NONE = if there's no lazy update to be performed.
+ */
+enum QueryType { ADD, SET, NONE };
 
-struct node {
-	ll val;
-	ll lzAdd;
-	ll lzSet;
-	node(){};
-} tree[maxN << 2];
+struct Query {
+	QueryType type = NONE;
+	ll val = 0;
+};
 
-#define lc p << 1
-#define rc (p << 1) + 1
+/**
+ * T = tree node type, which will be long long
+ * U = update type, which will be the Query struct
+ */
+template <class T, class U> class LazySegtree {
+  private:
+	const int sz;
+	vector<T> tree;  // tree[i] = sum of this node's range
+	vector<U> lazy;  // lazy[i] = lazy update for the range
+
+	/** @return result of joining two tree nodes together */
+	inline T comb(T a, T b) { return a + b; }
+
+	/** builds the segtree nodes */
+	void build(int v, int l, int r, const vector<T> &a) {
+		if (l == r) {
+			tree[v] = a[l];
+		} else {
+			int m = (l + r) / 2;
+			build(2 * v, l, m, a);
+			build(2 * v + 1, m + 1, r, a);
+			tree[v] = comb(tree[2 * v], tree[2 * v + 1]);
+		}
+	}
 
-inline void pushup(int p) {
-	tree[p].val = tree[lc].val + tree[rc].val;
-	return;
-}
+	/** applies lazy update to t[v], places update at lz[v] */
+	void apply(int v, int len, U x) {
+		if (x.type == ADD) {
+			// if lazy[v]'s type is NONE or ADD, then we add to the range
+			// otherwise, we add to our lazy set value
+			if (lazy[v].type != SET) {
+				lazy[v] = Query{ADD, lazy[v].val + x.val};
+			} else {
+				lazy[v] = Query{SET, lazy[v].val + x.val};
+			}
+			tree[v] += x.val * len;
+		} else if (query.type == SET) {
+			// lazy set overrides any previous update
+			tree[v] = x.val * len;
+			lazy[v] = x;
+		}
+	}
 
-void pushdown(int p, int l, int mid, int r) {
-	// lazy: range set
-	if (tree[p].lzSet != 0) {
-		tree[lc].lzSet = tree[rc].lzSet = tree[p].lzSet;
-		tree[lc].val = (mid - l + 1) * tree[p].lzSet;
-		tree[rc].val = (r - mid) * tree[p].lzSet;
-		tree[lc].lzAdd = tree[rc].lzAdd = 0;
-		tree[p].lzSet = 0;
-	} else if (tree[p].lzAdd != 0) {  // lazy: range add
-		if (tree[lc].lzSet == 0) tree[lc].lzAdd += tree[p].lzAdd;
-		else {
-			tree[lc].lzSet += tree[p].lzAdd;
-			tree[lc].lzAdd = 0;
+	/** pushes down lazy update to children of v */
+	void pushdown(int v, int l, int r) {
```

Comment:
why `pushdown` but `range_update` and `range_sum`

Reply:
oh shiiii i forgot to update it!